neurodsp/sim/multi.py

```diff
@@ -0,0 +1,166 @@
+"""Simulation functions that return multiple instances."""
+
+import numpy as np
+
+from neurodsp.utils.core import counter
+
+###################################################################################################
+###################################################################################################
+
+def sig_yielder(sim_func, sim_params, n_sims):
+    """Generator to yield simulated signals from a given simulation function and parameters.
+
+    Parameters
+    ----------
+    sim_func : callable
+        Function to create the simulated time series.
+    sim_params : dict
+        The parameters for the simulated signal, passed into `sim_func`.
+    n_sims : int, optional
+        Number of simulations to set as the max.
+        If None, creates an infinite generator.
+
+    Yields
+    ------
+    sig : 1d array
+        Simulated time series.
+    """
+
+    for _ in counter(n_sims):
+        yield sim_func(**sim_params)
+
+
+def sig_sampler(sim_func, sim_params, return_sim_params=False, n_sims=None):
+    """Generator to yield simulated signals from a parameter sampler.
+
+    Parameters
+    ----------
+    sim_func : callable
+        Function to create the simulated time series.
+    sim_params : iterable
+        The parameters for the simulated signal, passed into `sim_func`.
+    return_sim_params : bool, optional, default: False
+        Whether to yield the simulation parameters as well as the simulated time series.
+    n_sims : int, optional
+        Number of simulations to set as the max.
+        If None, creates an infinite generator.
+
+    Yields
+    ------
+    sig : 1d array
+        Simulated time series.
+    sample_params : dict
+        Simulation parameters for the yielded time series.
+    """
+
+    if len(sim_params) and n_sims and n_sims > len(sim_params):
+        msg = 'Cannot simulate the requested number of sims with the given parameters.'
+        raise ValueError(msg)
```

Comment:
Yeh - `sim_params` is a flexible input of an iterable, so it can be finite, in which case it effectively defines the number of sims, and `n_sims` is optional, and can define another way to define the number of sims (for example, if `sim_params` is infinite). 

Examples that show the different possibilities here:
```
from itertools import repeat
from neurodsp.sim.multi import sig_sampler
from neurodsp.sim.aperiodic import sim_powerlaw

params = [{'n_seconds' : 5, 'fs' : 250, 'exponent' : -1},
          {'n_seconds' : 5, 'fs' : 250, 'exponent' : -2}]
# This works - creates a generate for 2 signals
gen = sig_sampler(sim_powerlaw, params)

# This fails - n_sims requested beyond the length of params
gen = sig_sampler(sim_powerlaw, params, n_sims=3)

# This works - uses n_sims to define the length of infinite params def
gen = sig_sampler(sim_powerlaw, repeat(params[0]), n_sims=3)
```

I don't think this was super obvious from the docs / code (especially that the number of sims could be bound by the length of `sim_params` if `n_sims` is not set), so I've updated that a bit here (also to more properly support iterable inputs). More generally, this function isn't a user facing function - but the same logic applies to users passing things in through `sim_from_sampler`).